xfce4-dict.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH XFCE4-DICT "1" "October 2024" "xfce4-dict 0.8.7git-6f10536 (Xfce 4.18)" "User Commands"
.SH NAME
xfce4-dict \- manual page for xfce4-dict 0.8.7git-6f10536 (Xfce 4.18)
.SH DESCRIPTION
.SS "Usage:"
.IP
xfce4\-dict [OPTION?] [TEXT]
.SS "Help Options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Show help options
.TP
\fB\-\-help\-all\fR
Show all help options
.TP
\fB\-\-help\-gtk\fR
Show GTK+ Options
.SS "Application Options:"
.TP
\fB\-d\fR, \fB\-\-dict\fR
Search the given text using a Dict server(RFC 2229)
.TP
\fB\-w\fR, \fB\-\-web\fR
Search the given text using a web\-based search engine
.TP
\fB\-s\fR, \fB\-\-spell\fR
Check the given text with a spell checker
.TP
\fB\-t\fR, \fB\-\-text\-field\fR
Grab the focus on the text field in the panel
.TP
\fB\-i\fR, \fB\-\-ignore\-plugin\fR
Start stand\-alone application even if the panel plugin is loaded
.TP
\fB\-c\fR, \fB\-\-clipboard\fR
Grabs the PRIMARY selection content and uses it as search text
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Be verbose
.TP
\fB\-V\fR, \fB\-\-version\fR
Show version information
.TP
\fB\-\-display\fR=\fI\,DISPLAY\/\fR
X display to use
.SH COPYRIGHT
Copyright \(co 2006\-2023
.IP
Xfce Development Team
.PP
Please report bugs to <https://gitlab.xfce.org/apps/xfce4\-dict>.
.SH "SEE ALSO"
The full documentation for
.B xfce4-dict
is maintained as a Texinfo manual.  If the
.B info
and
.B xfce4-dict
programs are properly installed at your site, the command
.IP
.B info xfce4-dict
.PP
should give you access to the complete manual.


NEWS
0.8.8 (2024-10-01)
=====
- Avoid freeze when toggling search modes
- Drop --text-field (#18)
- Translation Updates:
  Portuguese (Brazil)

0.8.7 (2024-09-24)
=====
- panel-plugin: Drop submenu (#2)
- panel-plugin: Add submenus to toggle search mode (#2)
- panel-plugin: Reduce default text size
- panel-plugin: Restore function of the button in text entry
- Change log level (#17)
- prefs: Add radio buttons to correct group
- scan-build: Fix deadcode.DeadStores
- scan-build: Add false positive file
- I18n: Update po/LINGUAS list
- build: Use XDT_VERSION_INIT and get rid of configure.ac.in
- build: Switch from intltool to gettext
- Translation Updates:
  Albanian, Arabic, Armenian (Armenia), Basque, Belarusian, Bulgarian,
  Catalan, Chinese (China), Chinese (Taiwan), Croatian, Czech, Danish,
  Dutch, Eastern Armenian, English (Australia), English (United
  Kingdom), Estonian, Finnish, French, Galician, German, Greek, Hebrew,
  Hungarian, Icelandic, Indonesian, Italian, Japanese, Kazakh, Korean,
  Latvian, Lithuanian, Malay, Norwegian Bokmål, Occitan (post 1500),
  Persian (Iran), Polish, Portuguese, Portuguese (Brazil), Russian,
  Serbian, Slovak, Slovenian, Spanish, Swedish, Thai, Turkish,
  Ukrainian, Urdu, Urdu (Pakistan), Uyghur, Vietnamese

0.8.6 (2023-11-19)
=====
- Use getaddrinfo(3) to support IPv6 DICT servers
- Add icons at missing sizes, clean up SVG metadata


0.8.5 (2023-04-22)
=====
- build: Fix -Wunused-result warning
- Update `.gitignore`
- Update copyright year
- Use "logo-icon-name" instead of "logo" in about dialog
- Remove ENABLE_NLS ifdefs
- autoconf: Use AC_CONFIG_MACRO_DIRS
- build: Fix intltool lock file problem during make distcheck
- Update COPYING (Issue #7)
- Remove unnecessary code
- Use symbolic icons
- Escape markup in server information
- Replaces siginterrupt with sigaction
- Do not close window with ctrl+c (#5)
- Fix Text field size setting (#4)
- Fix compilation warnings
- autoconf: Some updates
- Remove "." at the end of Comment string
- Remove GSourceFunc casts
- Include stdlib.h to get atoi prototype (!2)
- Translation Updates:
  Albanian, Arabic, Armenian (Armenia), Basque, Belarusian, Bulgarian,
  Catalan, Chinese (China), Chinese (Taiwan), Croatian, Czech, Danish,
  Dutch, Eastern Armenian, English (Australia), English (United
  Kingdom), Estonian, Finnish, French, Galician, German, Greek, Hebrew,
  Hungarian, Icelandic, Indonesian, Italian, Japanese, Kazakh, Korean,
  Latvian, Lithuanian, Malay, Norwegian Bokmål, Occitan (post 1500),
  Persian (Iran), Polish, Portuguese, Portuguese (Brazil), Russian,
  Serbian, Slovak, Slovenian, Spanish, Swedish, Thai, Turkish,
  Ukrainian, Urdu, Urdu (Pakistan), Uyghur, Vietnamese

0.8.4
======
- Update URLs and use https where possible
- Fix font size in speedreader (#3)
- Added new icons and use reverse DNS file names
- Translation Updates:
  Albanian, Arabic, Armenian (Armenia), Basque, Belarusian, Bulgarian,
  Catalan, Chinese (China), Chinese (Taiwan), Croatian, Czech, Danish,
  Dutch, Eastern Armenian, English (Australia), English (United Kingdom),
  Estonian, Finnish, French, German, Greek, Hebrew, Hungarian, Icelandic,
  Indonesian, Italian, Japanese, Kazakh, Korean, Latvian, Lithuanian,
  Malay, Norwegian Bokmål, Occitan (post 1500), Polish, Portuguese,
  Russian, Serbian, Slovak, Slovenian, Spanish, Swedish, Thai, Turkish,
  Uighur, Ukrainian, Urdu, Urdu (Pakistan), Vietnamese

xfce4-dict-0.8.3
================
- Allow compilation with panel 4.15
- Show correct colors (Bug #15946)
- Avoid mnemonic conflict
- Fix g_type_class_add_private deprecation in recent GLib
- Replace GtkStock buttons
- Fix typo of dict.cc (Bug #15571)
- Fix parallel build (Bug #14963)
- Translation Updates: Albanian, Belarusian, Catalan, Chinese (China),
  Chinese (Taiwan), Croatian, Czech, Danish, Dutch, English (Australia),
  Finnish, French, Galician, German, Hebrew, Hungarian, Icelandic, Italian,
  Japanese, Kazakh, Korean, Lithuanian, Malay, Norwegian Bokmål, Polish,
  Portuguese, Portuguese (Brazil), Russian, Spanish, Thai, Turkish

xfce4-dict-0.8.2
================
- Avoid segfaults (Bug #14786)
- Only show web search link if we have a valid URL and in dict mode
- Escape link tooltip
- Minor visual improvements
- Translation updates: Albanian, Arabic, Asturian, Basque, Belarusian,
  Bulgarian, Catalan, Chinese (China), Chinese (Taiwan), Croatian, Czech,
  Danish, Dutch, English (United Kingdom), Finnish, French, Galician, German,
  Greek, Hebrew, Hungarian, Icelandic, Indonesian, Italian, Japanese, Korean,
  Latvian, Lithuanian, Norwegian Bokmål, Occitan, Polish, Portuguese,
  Portuguese (Brazil), Russian, Slovak, Slovenian, Spanish, Swedish, Turkish,
  Ukrainian, Urdu, Urdu (Pakistan), Uyghur, Vietnamese

xfce4-dict-0.8.1
================
- Bump gtk3 minimum version to 3.22
- Improve appearance of menu items with icons
- Append web search link to successful queries (Bug #12154)
- Translation updates: Albanian, Arabic, Asturian, Basque, Bulgarian, Catalan,
  Chinese (China), Chinese (Taiwan), Croatian, Czech, Danish, Dutch,
  English (Australia), English (United Kingdom), Finnish, French, Galician,
  German, Greek, Hebrew, Hungarian, Icelandic, Indonesian, Italian, Japanese,
  Kazakh, Korean, Latvian, Lithuanian, Malay, Norwegian Bokmål, Occitan, Polish,
  Portuguese, Portuguese (Brazil), Russian, Serbian, Slovak, Slovenian, Spanish,
  Swedish, Thai, Turkish, Ukrainian, Urdu, Urdu (Pakistan), Uyghur, Vietnamese

xfce4-dict-0.8.0
================
- Translation updates: Catalan, Chinese (China), Chinese (Taiwan), 
  Croatian, Danish, Dutch, Finnish, French, German, Greek, Hebrew, 
  Italian, Japanese, Kazakh, Korean, Malay, Norwegian Bokmål, Polish, 
  Portuguese, Russian, Serbian, Slovak, Slovenian, Spanish, Swedish, 
  Ukrainian

xfce4-dict-0.7.99
================
- Port to gtk+3/libxfce4ui-2/libxfce4panel-2.0
- Translation updates: French, Lithuanian, Portuguese (Brazilian)

xfce4-dict-0.7.2
================

src/xfce4-dict.c
/*  Copyright 2006-2012 Enrico Tröger <enrico(at)xfce(dot)org>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */


#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <gtk/gtk.h>
#include <string.h>
#include <stdlib.h>

#include <libxfce4util/libxfce4util.h>

#ifdef HAVE_LOCALE_H
# include <locale.h>
#endif

#include "libdict.h"
#include "popup_plugin.h"


static gboolean show_version = FALSE;
static gboolean ignore_plugin = FALSE;
static gboolean use_clipboard = FALSE;
static gboolean focus_panel_entry = FALSE;
static gboolean mode_dict = FALSE;
static gboolean mode_web = FALSE;
static gboolean mode_spell = FALSE;
static gboolean verbose_mode = FALSE;

static GOptionEntry cli_options[] =
{
	{ "dict", 'd', 0, G_OPTION_ARG_NONE, &mode_dict, N_("Search the given text using a Dict server(RFC 2229)"), NULL },
	{ "web", 'w', 0, G_OPTION_ARG_NONE, &mode_web, N_("Search the given text using a web-based search engine"), NULL },
	{ "spell", 's', 0, G_OPTION_ARG_NONE, &mode_spell, N_("Check the given text with a spell checker"), NULL },
	{ "ignore-plugin", 'i', 0, G_OPTION_ARG_NONE, &ignore_plugin, N_("Start stand-alone application even if the panel plugin is loaded"), NULL },
	{ "clipboard", 'c', 0, G_OPTION_ARG_NONE, &use_clipboard, N_("Grabs the PRIMARY selection content and uses it as search text"), NULL },
	{ "verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose_mode, N_("Be verbose"), NULL },
	{ "version", 'V', 0, G_OPTION_ARG_NONE, &show_version, N_("Show version information"), NULL },
	{ NULL, 0, 0, 0, NULL, NULL, NULL }
};


static gboolean main_quit(GtkWidget *widget, GdkEvent *event, DictData *dd)
{
	dict_gui_query_geometry(dd);
	dict_free_data(dd);
    gtk_main_quit();

    return FALSE;
}


static void close_button_clicked(GtkWidget *button, DictData *dd)
{
	main_quit(NULL, NULL, dd);
}


static void pref_dialog_activated(GtkMenuItem *menuitem, DictData *dd)
{
	GtkWidget *dlg;

	dlg = dict_prefs_dialog_show(gtk_widget_get_toplevel(GTK_WIDGET(menuitem)), dd);
	gtk_widget_show(dlg);
}


static gchar get_flags(void)
{
	gchar flags = 0;

	if (focus_panel_entry)
		flags |= DICT_FLAGS_FOCUS_PANEL_ENTRY;
	if (mode_dict)
		flags |= DICT_FLAGS_MODE_DICT;
	if (mode_web)
		flags |= DICT_FLAGS_MODE_WEB;
	if (mode_spell)
		flags |= DICT_FLAGS_MODE_SPELL;

	return flags;
}


static gchar *get_search_text(gint count, gchar **values)
{
	GString *str = g_string_sized_new(128);
	gint i;

	for (i = 1; i < count; i++)
	{
		g_string_append(str, values[i]);
		if (i < (count - 1))
			g_string_append_c(str, ' ');
	}
	return g_string_free(str, FALSE);
}


gint main(gint argc, gchar *argv[])
{
	DictData *dd;
	GOptionContext *context;
	gchar flags;
	gchar *search_text;

	xfce_textdomain(GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR, "UTF-8");

	context = g_option_context_new(_("[TEXT]"));
	g_option_context_add_main_entries(context, cli_options, GETTEXT_PACKAGE);
	g_option_group_set_translation_domain(g_option_context_get_main_group(context), GETTEXT_PACKAGE);
	g_option_context_add_group(context, gtk_get_option_group(FALSE));
	g_option_context_parse(context, &argc, &argv, NULL);

	gtk_init(&argc, &argv);
    gtk_window_set_default_icon_name("org.xfce.Dictionary");

	g_option_context_free(context);

	if (show_version)
	{
		g_print("%s %s (Xfce %s)\n\n", PACKAGE, PACKAGE_VERSION, xfce_version_string());
		g_print("%s\n", "Copyright (c) 2006-2023");
		g_print("\tXfce Development Team\n\n");
		g_print(_("Please report bugs to <%s>."), PACKAGE_BUGREPORT);
		g_print("\n");

		return EXIT_SUCCESS;
	}

	flags = get_flags();

	if (use_clipboard)
	{
		search_text = dict_get_clipboard_contents();
	}
	else
	{
		/* concatenate remaining command line arguments */
		search_text = get_search_text(argc, argv);
	}

	/* try to find an existing panel plugin and pop it up */
	if (! ignore_plugin && dict_find_panel_plugin(flags, search_text))
	{
		g_free(search_text);
		exit(0);
	}

	/* no plugin found, start stand-alone app */

	dd = dict_create_dictdata();
	dd->is_plugin = FALSE;
	dd->verbose_mode = verbose_mode;

	dict_read_rc_file(dd);

	/* set search mode from command line flags, if any */
	dd->mode_in_use = dict_set_search_mode_from_flags(dd->mode_in_use, flags);

	dict_gui_create_main_window(dd);

	g_signal_connect(dd->window, "delete-event", G_CALLBACK(main_quit), dd);
	g_signal_connect(dd->close_button, "clicked", G_CALLBACK(close_button_clicked), dd);
	/* file menu */
	g_signal_connect(dd->close_menu_item, "activate", G_CALLBACK(close_button_clicked), dd);
	g_signal_connect(dd->pref_menu_item, "activate", G_CALLBACK(pref_dialog_activated), dd);

	/* search text from command line options, if any */
	if (NZV(search_text))
	{
		gtk_entry_set_text(GTK_ENTRY(dd->main_entry), search_text);
		dict_search_word(dd, search_text);
	}
	else
		dict_gui_status_add(dd, _("Ready"));

	g_free(search_text);

	dict_acquire_dbus_name(dd);

	gtk_widget_show_all(dd->window);

	gtk_main();

	return 0;
}
